docs/ops/plugins.md

```diff
@@ -0,0 +1,91 @@
+---
+title: "Plugins"
+nav-id: plugins
+nav-parent_id: ops
+nav-pos: 16
+---
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+Plugins have been introduced in Flink 1.9 and facilitate a strict separation of code through restricted classloaders. Plugins cannot access classes from other plugins or from Flink that have not been specifically whitelisted. This strict isolation allows plugins to contain conflicting versions of the same library without the need to relocate classes or to converge to common versions.
+
+Currently, only file systems are pluggable but in the future, connectors, formats, and even user code should also be pluggable.
+
+* This will be replaced by the TOC
+{:toc}
+
+## Isolation and plugin structure
+
+Plugins reside in their own folders and can consist of several jars. The names of the plugin folders are arbitrary.
+
+```
+flink-dist
++-- conf
++-- lib
++-- ...
+\-- plugins
+    +-- s3
+    |   +-- aws-credential-provider.jar
+    |   \-- flink-s3-fs-hadoop.jar
+    \-- azure
+        \-- flink-azure-fs-hadoop.jar
+``` 
+
+Each plugin is loaded through it's own classloader and completely isolated from any other plugin. Hence, the `flink-s3-fs-hadoop` and `flink-azure-fs-hadoop` can depend on different conflicting library versions. There is no need to relocate any class during the creation of fat jars (shading).
+
+Plugins may access certain whitelisted packages from lib/. Most importantly, all needed service provider interfaces (SPI) are loaded through the system classloader, so that no two versions of `org.apache.flink.core.fs.FileSystem` exist at any given time, even if users accidentally bundle it in their fat jar. This singleton class requirement is strictly necessary, so that core Flink has an entry point into the plugin. Service classes are discovered through the `java.util.ServiceLoader`, so make sure to retain the service definitions in `META-INF/services` during shading.
```

Comment:
```suggestion
Plugins may access certain whitelisted packages from Flink's `lib/` folder. In particular, all necessary service provider interfaces (SPI) are loaded through the system classloader, so that no two versions of `org.apache.flink.core.fs.FileSystem` exist at any given time, even if users accidentally bundle it in their fat jar. This singleton class requirement is strictly necessary so that the Flink runtime has an entry point into the plugin. Service classes are discovered through the `java.util.ServiceLoader`, so make sure to retain the service definitions in `META-INF/services` during shading.
```